src/Integration.Vsix/SonarLintDaemon/LegacyInstallationCleanup.cs

```diff
@@ -0,0 +1,99 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using SonarLint.VisualStudio.Integration.Helpers;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    internal class LegacyInstallationCleanup
+    {
+        // See issue https://github.com/SonarSource/sonarlint-visualstudio/issues/470
+        // Previously the daemon and JVM binaries were downloaded the user's roaming profile
+        // e.g. c:\users\joebloggs\appdata\roaming\sonarlint-daemon-2.14.0.669-windows
+        // It's now downloaded elsewhere. This class attempts to clean up folders installed
+        // to the old location.
+
+        private readonly ILogger logger;
+        private readonly string legacyStoragePath;
+        private readonly IDirectory directory;
+
+        public static void CleanupDaemonFiles(ILogger logger)
+        {
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            new LegacyInstallationCleanup(logger,
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)),
+                new DirectoryWrapper()).Clean();
+        }
+
+        internal /* for testing */ LegacyInstallationCleanup(ILogger logger, string legacyStoragePath, IDirectory directory)
```

Comment:
I would not have done that this way but that works:
- I would have had this class responsible of getting the right directory (instead of finding directories under whatever directory provided) => - `legacyStoragePath` and - `directory`
- I would have created an interface for environment too.

Reply:
We'd still have to have provided an _IDirectory_ so we could mock returns from the file system.